be/src/service/backend_service.cpp

```diff
@@ -286,8 +286,121 @@ void _ingest_binlog(StorageEngine& engine, IngestBinlogArg* arg) {
+        status.to_thrift(&tstatus);
+        return;
+    }
+
+    // Step 6.3: get all segment index files
+    LOG(INFO) << "segment_index_file_urls.size=" << segment_index_file_urls.size() << ", "
+              << "segment_index_file_names.size=" << segment_index_file_names.size() << ", "
+              << "segment_index_file_sizes.size=" << segment_index_file_sizes.size();
+    DCHECK(segment_index_file_sizes.size() == segment_index_file_names.size());
+    DCHECK(segment_index_file_names.size() == segment_index_file_urls.size());
+    for (int64_t i = 0; i < segment_index_file_urls.size(); ++i) {
+        auto segment_index_file_size = segment_index_file_sizes[i];
+        auto get_segment_index_file_url = segment_index_file_urls[i];
+
+        uint64_t estimate_timeout =
+                segment_index_file_size / config::download_low_speed_limit_kbps / 1024;
+        if (estimate_timeout < config::download_low_speed_time) {
+            estimate_timeout = config::download_low_speed_time;
+        }
+
+        auto local_segment_index_path = segment_index_file_names[i];
+        LOG(INFO) << fmt::format("download segment index file from {} to {}",
+                                 get_segment_index_file_url, local_segment_index_path);
+        auto get_segment_index_file_cb = [&get_segment_index_file_url, &local_segment_index_path,
+                                          segment_index_file_size,
+                                          estimate_timeout](HttpClient* client) {
+            RETURN_IF_ERROR(client->init(get_segment_index_file_url));
+            client->set_timeout_ms(estimate_timeout * 1000);
+            RETURN_IF_ERROR(client->download(local_segment_index_path));
+
+            std::error_code ec;
+            // Check file length
+            uint64_t local_index_file_size =
+                    std::filesystem::file_size(local_segment_index_path, ec);
+            if (ec) {
+                LOG(WARNING) << "download index file error" << ec.message();
+                return Status::IOError("can't retrive file_size of {}, due to {}",
+                                       local_segment_index_path, ec.message());
+            }
+            if (local_index_file_size != segment_index_file_size) {
+                LOG(WARNING) << "download index file length error"
+                             << ", get_segment_index_file_url=" << get_segment_index_file_url
+                             << ", index_file_size=" << segment_index_file_size
+                             << ", local_index_file_size=" << local_index_file_size;
+                return Status::InternalError("downloaded index file size is not equal");
+            }
+            return io::global_local_filesystem()->permission(local_segment_index_path,
+                                                             io::LocalFileSystem::PERMS_OWNER_RW);
+        };
+
+        status = HttpClient::execute_with_retry(max_retry, 1, get_segment_index_file_cb);
```

Comment:
Is there gc logic for success files on failure?

Reply:
Added.